tests/unit/s2n_random_test.c

```diff
@@ -142,6 +146,110 @@ static S2N_RESULT s2n_basic_pattern_tests(S2N_RESULT (*s2n_get_random_data_cb)(s
     return S2N_RESULT_OK;
 }
 
+static void swap(size_t i, size_t j,
+    uint64_t range_results[NUMBER_OF_RANGE_FUNCTION_CALLS]) {
+    uint64_t temp = range_results[j];
+    range_results[j] = range_results[i];
+    range_results[i] = temp;
+}
+
+static void sort_array(uint64_t range_results[NUMBER_OF_RANGE_FUNCTION_CALLS]) {
+    for (size_t i = 1; i < NUMBER_OF_RANGE_FUNCTION_CALLS; i++) {
+        for (size_t j = i; j > 0; j--) {
+            if (range_results[j - 1] > range_results[j]) {
+                swap(j - 1, j, range_results);
+            }
+            else {
+                /* Invariant: subarray [0, i-1] is sorted */
+                continue;
+            }
+        }
+    }
+}
+
+static S2N_RESULT s2n_tests_get_range(S2N_RESULT (*s2n_get_range_cb)(int64_t bound, uint64_t *output))
+{
+    uint64_t range_results[NUMBER_OF_RANGE_FUNCTION_CALLS] = {0};
+    uint64_t current_bound;
+    uint64_t current_output;
+    /* The type of the `bound` parameter in s2n_public_random() is signed */
+    int64_t chosen_upper_bound;
+    struct s2n_blob upper_bound_blob = { .data = (void *) &chosen_upper_bound, .size = sizeof(chosen_upper_bound) };
+
+    /* 0 is not a legal upper bound */
+    current_bound = 0;
+    EXPECT_ERROR(s2n_get_range_cb(current_bound, &current_output));
+
+    /* For an upper bound of 1, 0 should be the only possible output */
+    current_bound = 1;
+    EXPECT_OK(s2n_get_range_cb(current_bound, &current_output));
+    EXPECT_EQUAL(current_output, 0);
+
+    /* For a upper bound of 2, 0 and 1 should be the only possible outputs */
+    current_bound = 1;
+    EXPECT_OK(s2n_get_range_cb(current_bound, &current_output));
+    EXPECT_TRUE((current_output == 0) || (current_output == 1));
+
+    /* Test NUMBER_OF_BOUNDS upper bounds. For each resulting range, draw
+     * NUMBER_OF_RANGE_FUNCTION_CALLS numbers from s2n_public_random() and
+     * verify the output. Set 2^30 * NUMBER_OF_RANGE_FUNCTION_CALLS as the
+     * minimal value for the upper bound.
+     */
+    int64_t minimal_upper_bound = (int64_t) 0x40000000 * (int64_t) NUMBER_OF_RANGE_FUNCTION_CALLS;
+    for (size_t bound_ctr = 0; bound_ctr < NUMBER_OF_BOUNDS; bound_ctr++) {
+
+UPPER_BOUND_TOO_SMALL:
+        EXPECT_OK(s2n_get_private_random_data(&upper_bound_blob));
+        if (chosen_upper_bound < minimal_upper_bound) {
+            goto UPPER_BOUND_TOO_SMALL;
```

Comment:
Is there a way to write this without using 'goto'? Anytime I try to reason about gotos I automatically collapse and fall to the floor and then I can't figure out where I left off.